src/rust/session/src/lib.rs

```diff
@@ -0,0 +1,244 @@
+    pub fn reregister(&mut self, poll: &Poll) -> Result<(), std::io::Error> {
+        let interest = self.readiness();
+        self.stream
+            .reregister(poll.registry(), self.token, interest)
+    }
+
+    /// Get the token which is used with the event loop
+    pub fn token(&self) -> Token {
+        self.token
+    }
+
+    /// Set the token which is used with the event loop
+    pub fn set_token(&mut self, token: Token) {
+        self.token = token;
+    }
+
+    /// Get the set of readiness events the session is waiting for
+    fn readiness(&self) -> Interest {
+        if self.write_buffer.is_empty() {
+            Interest::READABLE
+        } else {
+            Interest::READABLE | Interest::WRITABLE
+        }
+    }
+
+    /// Returns a boolean which indicates if the session is handshaking
+    pub fn is_handshaking(&self) -> bool {
+        self.stream.is_handshaking()
+    }
+
+    /// Drives the handshake for the session. A successful result indicates that
+    /// the session hadshake is completed successfully. The error result should
+    /// be checked to determine if the operation would block, resulted in some
+    /// unrecoverable error, or if the session was not in a handshaking state
+    /// when this was called.
+    pub fn do_handshake(&mut self) -> Result<(), std::io::Error> {
+        self.stream.do_handshake()
+    }
+
+    /// Closes the session and the underlying stream.
+    pub fn close(&mut self) {
+        self.stream.close();
+    }
+
+    /// Returns the number of bytes in the read buffer
+    pub fn read_pending(&self) -> usize {
+        self.read_buffer.len()
+    }
+
+    /// Returns the number of bytes in the write buffer
+    pub fn write_pending(&self) -> usize {
+        self.write_buffer.len()
+    }
+
+    pub fn write_capacity(&self) -> usize {
+        self.capacity.saturating_sub(self.write_pending())
+    }
+
+    /// Returns a reference to the internally buffered data.
+    ///
+    /// Unlike [`fill_buf`], this will not attempt to fill the buffer if it is
+    /// empty.
+    ///
+    /// [`fill_buf`]: BufRead::fill_buf
+    pub fn buffer(&self) -> &[u8] {
+        self.read_buffer.borrow()
+    }
+
+    pub fn peer_addr(&self) -> Result<SocketAddr, std::io::Error> {
+        self.stream.peer_addr()
+    }
+}
+
+impl Read for Session {
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize, std::io::Error> {
+        if self.read_buffer.is_empty() {
+            self.fill_buf()?;
+        }
+        let bytes = std::cmp::min(buf.len(), self.read_buffer.len());
+        let buffer: &[u8] = self.read_buffer.borrow();
+        buf[0..bytes].copy_from_slice(&buffer[0..bytes]);
+        self.consume(bytes);
+        Ok(bytes)
+    }
+}
+
+impl BufRead for Session {
+    fn fill_buf(&mut self) -> Result<&[u8], std::io::Error> {
+        increment_counter!(&Stat::SessionRecv);
+        let mut total_bytes = 0;
+        loop {
+            self.read_buffer.reserve(self.capacity);
```

Comment:
This might not be increasing capacity fast enough if we start from a low baseline `capacity` and the payload is large? Also, what's your thoughts on reserve size when the buffer length is non-zero? Should we stick to a fixed total size, or fixed extra capacity?

In the C buffer I started with doubling the total size of the buffer until a cutoff, at which point the incremental capacity is fixed. E.g.: `2K -> 4K -> 8K -> 16K -> 32K` (assuming this is the doubling cutoff) `-> 48K -> 64K -> ...`

Reply:
If we go with a 16KiB initial buffer size, should we forgo the doubling logic and just do linear growth?

Reply:
Per discussion - for now allowing simple doubling with a TODO around potentially implementing linear growth at a later point. Currently, linear growth would not be straightforward given the use of `Vec` as the underlying storage.